working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,294 @@
+
+A constructor can be made forwarding by writing a reference to a super-constructor (`super` or `super.id`) in a in place of a required positional parameter in the constructor parameter list *and* omitting the super-invocation in the initializer list. 
+
+The super-constructor entry is expanded to add a number of *forwarded parameters* to the forwarding constructor's parameter list. One positional forwarded parameter is added for each positional parameter of the referenced superconstructor, in order, at the position of the constructor reference in the forwarding constructor parameter list, and one named forwarded parameter is added for each named parameter of the superconstructor, with the following modifications:
+
+- If the forwarding constructor is also an initializing constructor, then the field-initializing parameters from that are added to the forwarding constructor before considering forwarding.
+- If the forwarding constructor declares a parameter (including initializing constructor parameters introduced above) with the same name as a required named superconstructor parameter, it is a *compile-time error*.
+- If the forwarding constructor declares a parameter (including initializing constructor parameters introduced above) with the same name as an optional named superconstructor parameter, then the superconstructor parameter is ignored and no corresponding parameter is introduced in the forwarding constructor
+- If the forwarding constructor declares a parameter (including initializing constructor parameters introduced above) with the same name as a positional superconstructor parameter, call it *originalName*, then, for documentation purposes, the corresponding forwarding constructor parameter uses the name *originalName_n* where *n* is a decimal integer representing for the smallest integer greater than zero which makes the name different from any parameter name declared by the forwarding constructor or by the superconstructor.
+- If the forwarding constructor declares any named parameters, including if it is an initializing constructor, then all optional positional superconstructor parameters are ignored.
+- If the forwarding constructor declares any positional parameters after the superconstructor reference entry, then all positional superconstructor-derived parameters are ignored.
+- Otherwise the forwarded parameters are optional if they are optional in the superconstructor, and if so, they have the same default value, if any.
+
+Then a super-invocation is added to the initializer list where each such forwarded parameter is forwarded as a corresponding superconstructor argument.
+
+#### Example
+
+The class `Fireman`above could be given an initializing constructor by writing:
+
+```dart
+  default Fireman(super);
```

Comment:
Can this have an initializer list?  What about a body?

Reply:
Yes to both.